### PR TITLE
Adding missing changes and leak fix

### PR DESCRIPTION
**Pull request by kpanwar**
Adding missing changes and leak fix
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-objc/pull/322%23discussion_r30822056%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-objc/pull/322%23discussion_r30822492%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-objc/pull/322%23discussion_r30824847%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-objc/pull/322%23discussion_r30858550%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-objc/pull/322%23discussion_r31074809%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-objc/pull/322%23discussion_r31074811%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-objc/commit/8ee88ff13a56918b1914d1cdf3d8a9efcac19949%23commitcomment-11313599%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-objc/commit/8ee88ff13a56918b1914d1cdf3d8a9efcac19949%23commitcomment-11318447%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-objc/commit/854a5191cd85533ebe9fd08227ab76c09baf8b0e%23commitcomment-11332377%22%2C%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-objc/commit/854a5191cd85533ebe9fd08227ab76c09baf8b0e%23commitcomment-11372509%22%5D%2C%20%22comments%22%3A%20%7B%22Commit%208ee88ff13a56918b1914d1cdf3d8a9efcac19949%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-objc/commit/8ee88ff13a56918b1914d1cdf3d8a9efcac19949%23commitcomment-11313599%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20sign%20off%20on%20this%20changes.%22%2C%20%22created_at%22%3A%20%222015-05-21T18%3A58%3A04Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/12160250%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cesarami%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-none%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-05-21T23%3A31%3A59Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7004783%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/RPangrle-MS%22%7D%7D%5D%2C%20%22title%22%3A%20%22Commit%208ee88ff%20comments%22%7D%2C%20%22Pull%208ee88ff13a56918b1914d1cdf3d8a9efcac19949%20ADALiOS/ADALiOS/ADAuthenticationContext.m%2034%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-objc/pull/322%23discussion_r30822056%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Having%20an%20-init%20method%20that%20takes%20no%20params%20and%20isn%27t%20named%20-init%20is%20weird.%22%2C%20%22created_at%22%3A%20%222015-05-21T16%3A55%3A14Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7004783%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/RPangrle-MS%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-none%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-05-26T20%3A16%3A04Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7004783%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/RPangrle-MS%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20ADALiOS/ADALiOS/ADAuthenticationContext.m%3AL1136-1170%22%7D%2C%20%22Pull%208ee88ff13a56918b1914d1cdf3d8a9efcac19949%20ADALiOS/ADALiOS/ADAuthenticationContext.m%2015%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-objc/pull/322%23discussion_r30824847%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22...%20this%20code%20is%20all%20iOS%20specific%2C%20meaning%20it%20should%20be%20using%20ARC%20and%20protected%20with%20ifdefs...%20so%20I%20don%27t%20think%20SAFE_ARC_RELEASE%20is%20the%20right%20answer%20here.%22%2C%20%22created_at%22%3A%20%222015-05-21T17%3A24%3A42Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7004783%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/RPangrle-MS%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-complete%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-05-21T23%3A31%3A59Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7004783%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/RPangrle-MS%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20ADALiOS/ADALiOS/ADAuthenticationContext.m%3AL124-131%22%7D%2C%20%22Pull%208ee88ff13a56918b1914d1cdf3d8a9efcac19949%20ADALiOS/ADALiOS/ADAuthenticationContext.m%2083%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-objc/pull/322%23discussion_r30822492%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22This%20chunk%20of%20code%20would%20be%20clearer%20as%20a%20separate%20function%20with%20early%20returns%20then%20with%20control%20blocks%20nested%205%20deep%22%2C%20%22created_at%22%3A%20%222015-05-21T16%3A59%3A40Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7004783%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/RPangrle-MS%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-complete%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-05-26T20%3A16%3A04Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7004783%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/RPangrle-MS%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20ADALiOS/ADALiOS/ADAuthenticationContext.m%3AL2239-2271%22%7D%2C%20%22Commit%20854a5191cd85533ebe9fd08227ab76c09baf8b0e%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/AzureAD/azure-activedirectory-library-for-objc/commit/854a5191cd85533ebe9fd08227ab76c09baf8b0e%23commitcomment-11332377%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20just%20tested%20this%20commit%20and%20the%20leaks%20seem%20to%20be%20gone.%20I%20sign%20off%20on%20this.%22%2C%20%22created_at%22%3A%20%222015-05-22T17%3A45%3A03Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/12160250%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cesarami%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%22%2C%20%22created_at%22%3A%20%222015-05-26T20%3A16%3A25Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7004783%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/RPangrle-MS%22%7D%7D%5D%2C%20%22title%22%3A%20%22Commit%20854a519%20comments%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [x] <a href='#crh-comment-Pull 8ee88ff13a56918b1914d1cdf3d8a9efcac19949 ADALiOS/ADALiOS/ADAuthenticationContext.m 34'></a> &nbsp;<b><a href='https://github.com/AzureAD/azure-activedirectory-library-for-objc/pull/322#discussion_r30822056'>File: ADALiOS/ADALiOS/ADAuthenticationContext.m:L1136-1170</a></b>
- <a href='https://github.com/RPangrle-MS'></a> Having an -init method that takes no params and isn't named -init is weird.
- [x] <a href='#crh-comment-Pull 8ee88ff13a56918b1914d1cdf3d8a9efcac19949 ADALiOS/ADALiOS/ADAuthenticationContext.m 83'></a> &nbsp;<b><a href='https://github.com/AzureAD/azure-activedirectory-library-for-objc/pull/322#discussion_r30822492'>File: ADALiOS/ADALiOS/ADAuthenticationContext.m:L2239-2271</a></b>
- <a href='https://github.com/RPangrle-MS'></a> This chunk of code would be clearer as a separate function with early returns then with control blocks nested 5 deep
- [x] <a href='#crh-comment-Pull 8ee88ff13a56918b1914d1cdf3d8a9efcac19949 ADALiOS/ADALiOS/ADAuthenticationContext.m 15'></a> &nbsp;<b><a href='https://github.com/AzureAD/azure-activedirectory-library-for-objc/pull/322#discussion_r30824847'>File: ADALiOS/ADALiOS/ADAuthenticationContext.m:L124-131</a></b>
- <a href='https://github.com/RPangrle-MS'></a> ... this code is all iOS specific, meaning it should be using ARC and protected with ifdefs... so I don't think SAFE_ARC_RELEASE is the right answer here.
- [x] <a href='#crh-comment-Commit 8ee88ff13a56918b1914d1cdf3d8a9efcac19949'></a> &nbsp;<b><a href='https://github.com/AzureAD/azure-activedirectory-library-for-objc/commit/8ee88ff13a56918b1914d1cdf3d8a9efcac19949#commitcomment-11313599'>Commit 8ee88ff comments</a></b>
- <a href='https://github.com/cesarami'></a> I sign off on this changes.
- [x] <a href='#crh-comment-Commit 854a5191cd85533ebe9fd08227ab76c09baf8b0e'></a> &nbsp;<b><a href='https://github.com/AzureAD/azure-activedirectory-library-for-objc/commit/854a5191cd85533ebe9fd08227ab76c09baf8b0e#commitcomment-11332377'>Commit 854a519 comments</a></b>
- <a href='https://github.com/cesarami'></a> I just tested this commit and the leaks seem to be gone. I sign off on this.

<a href='https://www.codereviewhub.com/AzureAD/azure-activedirectory-library-for-objc/pull/322?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/AzureAD/azure-activedirectory-library-for-objc/pull/322?approve=1'></a>&nbsp;<a href='https://github.com/AzureAD/azure-activedirectory-library-for-objc/pull/322'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by brandwe** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/AzureAD/azure-activedirectory-library-for-objc/pull/322'>Refresh</a>.<a href='#crh-update-none'></a>

File: ADALiOS/ADALiOS/ADAuthenticationContext.m [hunk 1]
```diff
@@ -1135,20 +1136,34 @@ -(void) handleBrokerFromWebiewResponse: (NSString*) urlString
...
+    if(![self respondsToUrl:[redirectUri absoluteString]])
+    {
+        error = [ADAuthenticationError errorFromAuthenticationError:AD_ERROR_INVALID_REDIRECT_URI
+                                                       protocolCode:nil
+                                                       errorDetails:ADRedirectUriInvalidError];
+        completionBlock([ADAuthenticationResult resultFromError:error]);
+        return;
+    }
+    
+    ADBrokerKeyHelper* brokerHelper = [[ADBrokerKeyHelper alloc] initHelper];
```
Comment (RPangrle-MS) [inline ADALiOS/ADALiOS/ADAuthenticationContext.m]: Having an -init method that takes no params and isn't named -init is weird.

File: ADALiOS/ADALiOS/ADAuthenticationContext.m [hunk 2]
```diff
@@ -2226,6 +2239,32 @@ - (BOOL) canUseBroker
 }
 
 
+- (BOOL) respondsToUrl:(NSString*)url
+{
+    BOOL schemeIsInPlist = NO; // find out if the sceme is in the plist file.
+    NSBundle* mainBundle = [NSBundle mainBundle];
+    NSArray* cfBundleURLTypes = [mainBundle objectForInfoDictionaryKey:@"CFBundleURLTypes"];
+    if ([cfBundleURLTypes isKindOfClass:[NSArray class]] && [cfBundleURLTypes lastObject]) {
```
Comment (RPangrle-MS) [inline ADALiOS/ADALiOS/ADAuthenticationContext.m]: This chunk of code would be clearer as a separate function with early returns then with control blocks nested 5 deep

File: ADALiOS/ADALiOS/ADAuthenticationContext.m [hunk 3]
```diff
@@ -124,6 +124,7 @@ + (void) load
                         [[UIApplication sharedApplication] setDelegate:(__bridge id)CFRetain((__bridge CFTypeRef)appDelegate)];
                     }
                     
+                    SAFE_ARC_RELEASE(observer);
```
Comment (RPangrle-MS) [inline ADALiOS/ADALiOS/ADAuthenticationContext.m]: ... this code is all iOS specific, meaning it should be using ARC and protected with ifdefs... so I don't think SAFE_ARC_RELEASE is the right answer here.

Comment (RPangrle-MS) [inline ADALiOS/ADALiOS/ADAuthenticationContext.m]: :+1:<a href='#crh-update-complete'></a>

File: ADALiOS/ADALiOS/ADAuthenticationContext.m (hunk 1 above)
Comment (RPangrle-MS) [inline ADALiOS/ADALiOS/ADAuthenticationContext.m]: :+1:<a href='#crh-update-none'></a>

File: ADALiOS/ADALiOS/ADAuthenticationContext.m (hunk 2 above)
Comment (RPangrle-MS) [inline ADALiOS/ADALiOS/ADAuthenticationContext.m]: :+1:<a href='#crh-update-complete'></a>

**Pull request merged**